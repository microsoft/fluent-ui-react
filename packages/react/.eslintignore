src/themes/teams/components/Icon/svg/ProcessedIcons
coverage/
dist/
lib/
node_modules/
