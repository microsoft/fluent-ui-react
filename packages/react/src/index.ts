//
// Theme
//
// work around api-extractor limitation
import { fontAwesome, teams, teamsDark, teamsHighContrast } from './themes'

export const themes = { fontAwesome, teams, teamsDark, teamsHighContrast }
export * from './themes/types'
export * from './themes/colorUtils'

//
// Teams theme
//
export * from './themes/teams/types'

//
// Components
//
export * from './components/Accordion/Accordion'
export { default as Accordion } from './components/Accordion/Accordion'
export * from './components/Accordion/AccordionTitle'
export { default as AccordionTitle } from './components/Accordion/AccordionTitle'
export * from './components/Accordion/AccordionContent'
export { default as AccordionContent } from './components/Accordion/AccordionContent'

export * from './components/Alert/Alert'
export { default as Alert } from './components/Alert/Alert'

export * from './components/Attachment/Attachment'
export { default as Attachment } from './components/Attachment/Attachment'

export * from './components/Avatar/Avatar'
export { default as Avatar } from './components/Avatar/Avatar'

export * from './components/Box/Box'
export { default as Box } from './components/Box/Box'

export * from './components/Button/Button'
export { default as Button } from './components/Button/Button'
export * from './components/Button/ButtonGroup'
export { default as ButtonGroup } from './components/Button/ButtonGroup'

export * from './components/Chat/Chat'
export { default as Chat } from './components/Chat/Chat'
export * from './components/Chat/ChatItem'
export { default as ChatItem } from './components/Chat/ChatItem'
export * from './components/Chat/ChatMessage'
export { default as ChatMessage } from './components/Chat/ChatMessage'

export * from './components/Checkbox/Checkbox'
export { default as Checkbox } from './components/Checkbox/Checkbox'

export * from './components/Debug/Debug'
export { default as Debug } from './components/Debug/Debug'

export * from './components/Design/Design'
export { default as Design } from './components/Design/Design'

export * from './components/MenuButton/MenuButton'
export { default as MenuButton } from './components/MenuButton/MenuButton'

export * from './components/Divider/Divider'
export { default as Divider } from './components/Divider/Divider'

export * from './components/Dialog/Dialog'
export { default as Dialog } from './components/Dialog/Dialog'

export * from './components/Dialog/DialogFooter'
export { default as DialogFooter } from './components/Dialog/DialogFooter'

export * from './components/Dropdown/Dropdown'
export { default as Dropdown } from './components/Dropdown/Dropdown'
export * from './components/Dropdown/DropdownItem'
export { default as DropdownItem } from './components/Dropdown/DropdownItem'
export * from './components/Dropdown/DropdownSelectedItem'
export { default as DropdownSelectedItem } from './components/Dropdown/DropdownSelectedItem'
export * from './components/Dropdown/DropdownSearchInput'
export { default as DropdownSearchInput } from './components/Dropdown/DropdownSearchInput'

export * from './components/Embed/Embed'
export { default as Embed } from './components/Embed/Embed'

export * from './components/Flex/Flex'
export { default as Flex } from './components/Flex/Flex'
export * from './components/Flex/FlexItem'
export { default as FlexItem } from './components/Flex/FlexItem'

export * from './components/Form/Form'
export { default as Form } from './components/Form/Form'
export * from './components/Form/FormField'
export { default as FormField } from './components/Form/FormField'

export * from './components/Grid/Grid'
export { default as Grid } from './components/Grid/Grid'

export * from './components/Header/Header'
export { default as Header } from './components/Header/Header'
export * from './components/Header/HeaderDescription'
export { default as HeaderDescription } from './components/Header/HeaderDescription'

export * from './components/Icon/Icon'
export { default as Icon } from './components/Icon/Icon'

export * from './components/Image/Image'
export { default as Image } from './components/Image/Image'

export * from './components/Input/Input'
export { default as Input } from './components/Input/Input'

export * from './components/ItemLayout/ItemLayout'
export { default as ItemLayout } from './components/ItemLayout/ItemLayout'

export * from './components/Label/Label'
export { default as Label } from './components/Label/Label'

export * from './components/Loader/Loader'
export { default as Loader } from './components/Loader/Loader'

export * from './components/Layout/Layout'
export { default as Layout } from './components/Layout/Layout'

export * from './components/List/List'
export { default as List } from './components/List/List'
export * from './components/List/ListItem'
export { default as ListItem } from './components/List/ListItem'

export * from './components/Menu/Menu'
export { default as Menu } from './components/Menu/Menu'
export * from './components/Menu/MenuItem'
export { default as MenuItem } from './components/Menu/MenuItem'
export * from './components/Menu/MenuDivider'
export { default as MenuDivider } from './components/Menu/MenuDivider'

export * from './components/Popup/Popup'
export { default as Popup } from './components/Popup/Popup'
export * from './components/Popup/PopupContent'
export { default as PopupContent } from './components/Popup/PopupContent'

export * from './components/Portal/Portal'
export { default as Portal } from './components/Portal/Portal'

export * from './components/Provider/Provider'
export { default as Provider } from './components/Provider/Provider'
export * from './components/Provider/ProviderConsumer'
export { default as ProviderConsumer } from './components/Provider/ProviderConsumer'

export * from './components/RadioGroup/RadioGroup'
export { default as RadioGroup } from './components/RadioGroup/RadioGroup'
export * from './components/RadioGroup/RadioGroupItem'
export { default as RadioGroupItem } from './components/RadioGroup/RadioGroupItem'

export * from './components/Segment/Segment'
export { default as Segment } from './components/Segment/Segment'

export * from './components/Slider/Slider'
export { default as Slider } from './components/Slider/Slider'

export * from './components/Status/Status'
export { default as Status } from './components/Status/Status'

export * from './components/Text/Text'
export { default as Text } from './components/Text/Text'

export * from './components/Animation/Animation'
export { default as Animation } from './components/Animation/Animation'

export * from './components/TextArea/TextArea'
export { default as TextArea } from './components/TextArea/TextArea'

export * from './components/Toolbar/Toolbar'
export { default as Toolbar } from './components/Toolbar/Toolbar'
export * from './components/Toolbar/ToolbarCustomItem'
export { default as ToolbarCustomItem } from './components/Toolbar/ToolbarCustomItem'
export * from './components/Toolbar/ToolbarDivider'
export { default as ToolbarDivider } from './components/Toolbar/ToolbarDivider'
export * from './components/Toolbar/ToolbarItem'
export { default as ToolbarItem } from './components/Toolbar/ToolbarItem'
export * from './components/Toolbar/ToolbarMenu'
export { default as ToolbarMenu } from './components/Toolbar/ToolbarMenu'
export * from './components/Toolbar/ToolbarMenuDivider'
export { default as ToolbarMenuDivider } from './components/Toolbar/ToolbarMenuDivider'
export * from './components/Toolbar/ToolbarMenuItem'
export { default as ToolbarMenuItem } from './components/Toolbar/ToolbarMenuItem'
export * from './components/Toolbar/ToolbarMenuRadioGroup'
export { default as ToolbarMenuRadioGroup } from './components/Toolbar/ToolbarMenuRadioGroup'
export * from './components/Toolbar/ToolbarRadioGroup'
export { default as ToolbarRadioGroup } from './components/Toolbar/ToolbarRadioGroup'

export * from './components/HierarchicalTree/HierarchicalTree'
export { default as HierarchicalTree } from './components/HierarchicalTree/HierarchicalTree'
export * from './components/HierarchicalTree/HierarchicalTreeItem'
export { default as HierarchicalTreeItem } from './components/HierarchicalTree/HierarchicalTreeItem'
export * from './components/HierarchicalTree/HierarchicalTreeTitle'
export { default as HierarchicalTreeTitle } from './components/HierarchicalTree/HierarchicalTreeTitle'

export * from './components/Tree/Tree'
export { default as Tree } from './components/Tree/Tree'
export * from './components/Tree/TreeItem'
export { default as TreeItem } from './components/Tree/TreeItem'
export * from './components/Tree/TreeTitle'
export { default as TreeTitle } from './components/Tree/TreeTitle'

export * from './components/Reaction/Reaction'
export { default as Reaction } from './components/Reaction/Reaction'
export * from './components/Reaction/ReactionGroup'
export { default as ReactionGroup } from './components/Reaction/ReactionGroup'

export * from './components/SplitButton/SplitButton'
export { default as SplitButton } from './components/SplitButton/SplitButton'

export * from './components/Video/Video'
export { default as Video } from './components/Video/Video'

export * from './components/Tooltip/Tooltip'
export { default as Tooltip } from './components/Tooltip/Tooltip'
export * from './components/Tooltip/TooltipContent'
export { default as TooltipContent } from './components/Tooltip/TooltipContent'

export * from './components/Carousel/Carousel'
export { default as Carousel } from './components/Carousel/Carousel'
export * from './components/Carousel/CarouselItem'
export { default as CarouselItem } from './components/Carousel/CarouselItem'
export * from './components/Carousel/CarouselNavigation'
export { default as CarouselNavigation } from './components/Carousel/CarouselNavigation'
export * from './components/Carousel/CarouselNavigationItem'
export { default as CarouselNavigationItem } from './components/Carousel/CarouselNavigationItem'

export * from './components/Table/Table'
export { default as Table } from './components/Table/Table'
export * from './components/Table/TableRow'
export { default as TableRow } from './components/Table/TableRow'
export * from './components/Table/TableCell'
export { default as TableCell } from './components/Table/TableCell'

//
// Utilities
//
export * from './utils'
export * from './types'
export { Popper as UNSTABLE_Popper } from './utils/positioner'
export * from './utils/positioner/types'

//
// FocusZone
//
import {
  getFirstTabbable,
  getLastTabbable,
  getFirstFocusable,
  getLastFocusable,
  getNextElement,
  getPreviousElement,
  focusAsync,
} from '@fluentui/react-bindings'

export const FocusZoneUtilities = {
  getFirstTabbable,
  getLastTabbable,
  getFirstFocusable,
  getLastFocusable,
  getNextElement,
  getPreviousElement,
  focusAsync,
}

export * from '@fluentui/accessibility'
export * from '@fluentui/react-component-ref'
export * from '@fluentui/react-bindings'
export * from '@fluentui/styles'
