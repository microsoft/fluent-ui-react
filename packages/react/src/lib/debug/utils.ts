export const containsSubstring = (arg: any, substring: string) => {
  return typeof arg === 'string' && arg.indexOf(substring) >= 0
}

export const getLastOf = function<Item>(array: Item[], fallback: Item) {
  return array.length > 0 ? array[array.length - 1] : fallback
}

export const isNotNullOrEmpty = object => {
  if (object == null) {
    return false
  }

  return Object.keys(object).length > 0
}

export const isNotEmptyObjectsArray = array => {
  if (array == null) {
    return false
  }

  return array.some(isNotNullOrEmpty)
}

export const deepPickBy = (
  object: any,
  predicate: (propName: string, propValue: any) => boolean,
  seenObjects = [],
) => {
  if (!object) {
    return {}
  }

  if (seenObjects.some(seenObject => seenObject === object)) {
    throw new Error(`Circular dependency detected.`)
  }

  return Object.keys(object).reduce((acc, currentPropName) => {
    if (predicate(currentPropName, object[currentPropName])) {
      return { ...acc, [currentPropName]: object[currentPropName] }
    }

    if (typeof object[currentPropName] === 'object') {
      const deepPickResult = deepPickBy(object[currentPropName], predicate, [
        ...seenObjects,
        object,
      ])

      if (isNotNullOrEmpty(deepPickResult)) {
        return { ...acc, [currentPropName]: deepPickResult }
      }
    }

    return acc
  }, {})
}

export const deepPick = (object, propName) =>
  deepPickBy(object, currentPropName => propName === currentPropName)
