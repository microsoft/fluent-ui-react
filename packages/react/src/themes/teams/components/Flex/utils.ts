export const toFlexAlignment = propValue => {
  const trimmedValue = propValue.trim()

  if (trimmedValue === 'start' || trimmedValue === 'end') {
    return `flex-${trimmedValue}`
  }

  return trimmedValue
}

export const toFlexItemSizeValues = sizeValue => {
  return {
    flexBasis: sizeValue,
  }
}
