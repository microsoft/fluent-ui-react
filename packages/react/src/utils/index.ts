export { default as applyAccessibilityKeyHandlers } from './applyAccessibilityKeyHandlers'
export { default as AutoControlledComponent } from './AutoControlledComponent'
export { default as childrenExist } from './childrenExist'
export { default as UIComponent } from './UIComponent'
export { createRenderer, felaRenderer } from './felaRenderer'
export { default as toCompactArray } from './toCompactArray'
export { default as rtlTextContainer } from './rtlTextContainer'
export { default as stringLiteralsArray } from './stringLiteralsArray'
export { default as getOrGenerateIdFromShorthand } from './getOrGenerateIdFromShorthand'

export * from './factories'
export { default as constants } from './constants'
export { default as mergeThemes } from './mergeThemes'
export { default as mergeProviderContexts } from './mergeProviderContexts'

export * from './renderComponent'
export { default as renderComponent } from './renderComponent'

export { default as getElementProp } from './getElementProp'

export {
  htmlImageProps,
  htmlInputAttrs,
  htmlInputEvents,
  htmlInputProps,
  partitionHTMLProps,
} from './htmlPropsUtils'

export { default as isBrowser } from './isBrowser'
export { default as doesNodeContainClick } from './doesNodeContainClick'

export { pxToRem } from './fontSizeUtility'
export { default as createAnimationStyles } from './createAnimationStyles'
export { default as createComponent } from './createComponent'
export { getKindProp } from './getKindProp'
export * from './whatInput'

export * from './commonPropInterfaces'
// work around api-extractor limitation
import {
  CreateCommonConfig as CreateCommonConfigLocal,
  createCommon as createCommonLocal,
} from './commonPropTypes'

export module commonPropTypes {
  export type CreateCommonConfig = CreateCommonConfigLocal
  export const createCommon = createCommonLocal
}

export { default as withDebugId } from './withDebugId'
export { default as Telemetry } from './Telemetry'
