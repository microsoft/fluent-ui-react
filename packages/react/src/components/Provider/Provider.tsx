import { IStyle } from 'fela'
import { render } from 'fela-dom'
import * as _ from 'lodash'
import * as PropTypes from 'prop-types'
import * as React from 'react'
import { RendererProvider, ThemeProvider } from 'react-fela'

import { felaRenderer as felaLtrRenderer, isBrowser, mergeThemes } from '../../lib'
import {
  ThemePrepared,
  ThemeInput,
  StaticStyleObject,
  StaticStyle,
  StaticStyleFunction,
  FontFace,
} from '../../themes/types'
import ProviderConsumer from './ProviderConsumer'
import { mergeSiteVariables } from '../../lib/mergeThemes'

export interface ProviderProps {
  theme: ThemeInput
  children: React.ReactNode
}

/**
 * The Provider passes the CSS in JS renderer and theme to your components.
 */
class Provider extends React.Component<ProviderProps> {
  staticStylesRendered: boolean = false

  static propTypes = {
    theme: PropTypes.shape({
      siteVariables: PropTypes.object,
      componentVariables: PropTypes.object,
      componentStyles: PropTypes.object,
      rtl: PropTypes.bool,
      fontFaces: PropTypes.arrayOf(
        PropTypes.shape({
          name: PropTypes.string,
          paths: PropTypes.arrayOf(PropTypes.string),
          style: PropTypes.shape({
            fontStretch: PropTypes.string,
            fontStyle: PropTypes.string,
            fontVariant: PropTypes.string,
            fontWeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
            localAlias: PropTypes.string,
            unicodeRange: PropTypes.string,
          }),
        }),
      ),
      staticStyles: PropTypes.arrayOf(
        PropTypes.oneOfType([PropTypes.string, PropTypes.object, PropTypes.func]),
      ),
      animations: PropTypes.object,
    }),
    children: PropTypes.element.isRequired,
  }

  static Consumer = ProviderConsumer

  renderStaticStyles = (mergedTheme: ThemePrepared) => {
    // RTL WARNING
    // This function sets static styles which are global and renderer agnostic
    // With current implementation, static styles cannot differ between LTR and RTL
    // @see http://fela.js.org/docs/advanced/StaticStyle.html for details

    const { siteVariables } = mergedTheme
    const { staticStyles } = this.props.theme

    if (!staticStyles) return

    const renderObject = (object: StaticStyleObject) => {
      _.forEach(object, (style, selector) => {
        felaLtrRenderer.renderStatic(style as IStyle, selector)
      })
    }

    staticStyles.forEach((staticStyle: StaticStyle) => {
      if (typeof staticStyle === 'string') {
        felaLtrRenderer.renderStatic(staticStyle)
      } else if (_.isPlainObject(staticStyle)) {
        renderObject(staticStyle as StaticStyleObject)
      } else if (_.isFunction(staticStyle)) {
        const preparedSiteVariables = mergeSiteVariables(siteVariables)
        renderObject((staticStyle as StaticStyleFunction)(preparedSiteVariables))
      } else {
        throw new Error(
          `staticStyles array must contain CSS strings, style objects, or style functions, got: ${typeof staticStyle}`,
        )
      }
    })
  }

  renderFontFaces = () => {
    // RTL WARNING
    // This function sets static styles which are global and renderer agnostic
    // With current implementation, static styles cannot differ between LTR and RTL
    // @see http://fela.js.org/docs/advanced/StaticStyle.html for details

    const { fontFaces } = this.props.theme

    if (!fontFaces) return

    const renderFontObject = (font: FontFace) => {
      if (!_.isPlainObject(font)) {
        throw new Error(`fontFaces must be objects, got: ${typeof font}`)
      }
      felaLtrRenderer.renderFont(font.name, font.paths, font.style)
    }

    fontFaces.forEach((font: FontFace) => {
      renderFontObject(font)
    })
  }

  componentDidMount() {
    this.renderFontFaces()
  }

  render() {
    const { theme, children } = this.props

    // rehydration disabled to avoid leaking styles between renderers
    // https://github.com/rofrischmann/fela/blob/master/docs/api/fela-dom/rehydrate.md
    return (
      <ProviderConsumer
        render={(incomingTheme: ThemePrepared) => {
          const outgoingTheme: ThemePrepared = mergeThemes(incomingTheme, theme)

          // Heads up!
          // We should call render() to ensure that a subscription for DOM updates was created
          // https://github.com/stardust-ui/react/issues/581
          if (isBrowser()) render(outgoingTheme.renderer)
          this.renderStaticStylesOnce(outgoingTheme)

          return (
            <RendererProvider renderer={outgoingTheme.renderer} {...{ rehydrate: false }}>
              <ThemeProvider theme={outgoingTheme}>{children}</ThemeProvider>
            </RendererProvider>
          )
        }}
      />
    )
  }

  renderStaticStylesOnce = (mergedTheme: ThemePrepared) => {
    const { staticStyles } = this.props.theme
    if (!this.staticStylesRendered && staticStyles) {
      this.renderStaticStyles(mergedTheme)
      this.staticStylesRendered = true
    }
  }
}

export default Provider
