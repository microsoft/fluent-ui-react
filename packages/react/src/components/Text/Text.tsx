import * as PropTypes from 'prop-types'
import * as React from 'react'

import {
  childrenExist,
  createShorthandFactory,
  customPropTypes,
  UIComponent,
  UIComponentProps,
  ContentComponentProps,
  ChildrenComponentProps,
  commonPropTypes,
  ColorComponentProps,
  rtlTextContainer,
  SizeValue,
} from '../../lib'
import { Accessibility } from '../../lib/accessibility/types'
import { defaultBehavior } from '../../lib/accessibility'
import { ReactProps } from '../../types'

export interface TextProps
  extends UIComponentProps,
    ContentComponentProps,
    ChildrenComponentProps,
    ColorComponentProps {
  /**
   * Accessibility behavior if overridden by the user.
   * @default defaultBehavior
   */
  accessibility?: Accessibility

  /** At mentions can be formatted to draw users' attention. Mentions for "me" can be formatted to appear differently. */
  atMention?: boolean | 'me'

  /** Set as disabled Text component */
  disabled?: boolean

  /** Set as error Text component */
  error?: boolean

  /** The text can appear more important and draw user's attention */
  important?: boolean

  /** The size for the Text component */
  size?: SizeValue

  /** The weight for the Text component */
  weight?: 'light' | 'semilight' | 'regular' | 'semibold' | 'bold'

  /** Set as success Text component */
  success?: boolean

  /** The text can signify a temporary state */
  temporary?: boolean

  /** Set as timestamp Text component */
  timestamp?: boolean

  /** Truncates text as needed */
  truncated?: boolean
}

/**
 * A Text component formats occurrences of text consistently.
 * @accessibility
 * Text is how people read the content on your website.
 * Ensure that a contrast ratio of at least 4.5:1 exists between text and the background behind the text.
 *
 * To ensure that RTL mode will be properly handled for provided 'content' value, ensure that either:
 * - 'content' is provided as plain string (then dir="auto" attribute will be applied automatically)
 * - for other 'content' value types (i.e. that use elements inside) ensure that dir="auto" attribute is applied for all places in content where necessary
 */
class Text extends UIComponent<ReactProps<TextProps>, any> {
  static create: Function

  static className = 'ui-text'

  static displayName = 'Text'

  static propTypes = {
    ...commonPropTypes.createCommon({ color: true }),
    atMention: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['me'])]),
    disabled: PropTypes.bool,
    error: PropTypes.bool,
    important: PropTypes.bool,
    size: customPropTypes.size,
    weight: PropTypes.oneOf(['light', 'semilight', 'regular', 'semibold', 'bold']),
    success: PropTypes.bool,
    temporary: PropTypes.bool,
    timestamp: PropTypes.bool,
    truncated: PropTypes.bool,
  }

  static defaultProps = {
    accessibility: defaultBehavior,
    as: 'span',
  }

  renderComponent({ accessibility, ElementType, classes, unhandledProps }): React.ReactNode {
    const { children, content } = this.props

    return (
      <ElementType
        className={classes.root}
        {...rtlTextContainer.getAttributes({ forElements: [children, content] })}
        {...accessibility.attributes.root}
        {...unhandledProps}
      >
        {childrenExist(children) ? children : content}
      </ElementType>
    )
  }
}

Text.create = createShorthandFactory(Text, 'content')

export default Text
