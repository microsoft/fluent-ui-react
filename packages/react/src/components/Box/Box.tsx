import * as React from 'react'
import {
  childrenExist,
  createShorthandFactory,
  UIComponentProps,
  ContentComponentProps,
  ChildrenComponentProps,
  commonPropTypes,
  rtlTextContainer,
} from '../../lib'
import createComponent, { CreateComponentReturnType } from '../../lib/createComponent'
import { ReactProps } from '../../../types/utils'

export interface BoxProps
  extends UIComponentProps<BoxProps>,
    ContentComponentProps,
    ChildrenComponentProps {}

/**
 * A Box is a basic component (no default styles)
 */
const Box: CreateComponentReturnType<ReactProps<BoxProps>> = createComponent<BoxProps>({
  displayName: 'Box',

  className: 'ui-box',

  propTypes: {
    ...commonPropTypes.createCommon(),
  },

  render(config, props) {
    const { ElementType, classes, unhandledProps } = config
    const { children, content } = props

    return (
      <ElementType
        {...rtlTextContainer.getAttributes({ forElements: [children, content] })}
        {...unhandledProps}
        className={classes.root}
      >
        {childrenExist(children) ? children : content}
      </ElementType>
    )
  },
})

Box.create = createShorthandFactory(Box)

export default Box
