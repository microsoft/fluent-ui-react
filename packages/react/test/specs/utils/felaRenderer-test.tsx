import * as React from 'react'
import { createSnapshot } from 'jest-react-fela'
import { EmptyThemeProvider } from 'test/utils'
import Box from 'src/components/Box/Box'
import Provider from 'src/components/Provider/Provider'
import Text from 'src/components/Text/Text'
import { felaRenderer } from 'src/utils'

describe('felaRenderer', () => {
  test('basic styles are rendered', () => {
    const snapshot = createSnapshot(
      <EmptyThemeProvider>
        <Box styles={{ color: 'red' }} />
      </EmptyThemeProvider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })

  test('CSS fallback values are rendered', () => {
    const snapshot = createSnapshot(
      <EmptyThemeProvider>
        <Box styles={{ color: ['red', 'blue'] as any }} />
      </EmptyThemeProvider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })

  test('keyframe colors are rendered', () => {
    const spinner = {
      keyframe: ({ fromColor, toColor }) => ({
        from: {
          color: fromColor,
        },
        to: {
          color: toColor,
        },
      }),
      keyframeParams: {
        fromColor: 'red',
        toColor: 'blue',
      },
      duration: '5s',
    }

    const snapshot = createSnapshot(
      <Provider theme={{ animations: { spinner } }}>
        <Box animation="spinner" />
      </Provider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })

  test('array returned by keyframe results in CSS fallback values', () => {
    const steps = ['0%', '100%']

    const spinner = {
      keyframe: ({ steps }) => {
        const obj = {}
        steps.forEach((step: string, index) => {
          ;(obj as any)[step] = { color: ['blue', 'red', 'yellow'] }
        })
        return obj
      },
      keyframeParams: { steps },
    }

    const snapshot = createSnapshot(
      <Provider theme={{ animations: { spinner } }}>
        <Box animation="spinner" />
      </Provider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })

  test('animations are not applied if animations are disabled', () => {
    const spinner = {
      keyframe: {
        from: {
          transform: 'rotate(0deg)',
        },
        to: {
          transform: 'rotate(360deg)',
        },
      },
      duration: '5s',
      iterationCount: 'infinite',
    }

    const snapshot = createSnapshot(
      <Provider disableAnimations theme={{ animations: { spinner } }}>
        <Box animation="spinner" />
      </Provider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })

  test('marginLeft is rendered into marginRight due to RTL', () => {
    const snapshot = createSnapshot(
      <Provider rtl>
        <Text content="Hello" styles={{ marginLeft: '10px' }} />
      </Provider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })

  test('marginLeft is rendered into marginLeft due to RTL with `noFlip`', () => {
    const snapshot = createSnapshot(
      <Provider rtl>
        <Text content="Hello" styles={{ marginLeft: '10px /* @noflip */' }} />
      </Provider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })

  test('styles are expanded to longhand values', () => {
    const snapshot = createSnapshot(
      <EmptyThemeProvider>
        <Box
          styles={{
            borderStyle: 'solid',
            // spaces in color value are important
            borderColor: 'rgba(51,204, 51, 1) rgba(51,0,204, 1)',
          }}
        />
      </EmptyThemeProvider>,
      {},
      felaRenderer,
    )
    expect(snapshot).toMatchSnapshot()
  })
})
