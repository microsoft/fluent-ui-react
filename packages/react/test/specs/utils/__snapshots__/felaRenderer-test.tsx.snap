// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`felaRenderer CSS fallback values are rendered 1`] = `
".a {
  color: red;
  color: blue;
}


<div className=ui-box a />;
"
`;

exports[`felaRenderer animations are not applied if animations are disabled 1`] = `
"

<div className=ui-provider dir=ltr>
  <div className=ui-box ui-animation />
</div>;
"
`;

exports[`felaRenderer array returned by keyframe results in CSS fallback values 1`] = `
"@-webkit-keyframes k1 {
  0% {
    color: blue;
    color: red;
    color: yellow;
  }
  100% {
    color: blue;
    color: red;
    color: yellow;
  }
}
@-moz-keyframes k1 {
  0% {
    color: blue;
    color: red;
    color: yellow;
  }
  100% {
    color: blue;
    color: red;
    color: yellow;
  }
}
@keyframes k1 {
  0% {
    color: blue;
    color: red;
    color: yellow;
  }
  100% {
    color: blue;
    color: red;
    color: yellow;
  }
}
.a {
  animation-name: k1;
}


<div className=ui-provider dir=ltr>
  <div className=ui-box ui-animation a />
</div>;
"
`;

exports[`felaRenderer basic styles are rendered 1`] = `
".a {
  color: red;
}


<div className=ui-box a />;
"
`;

exports[`felaRenderer keyframe colors are rendered 1`] = `
"@-webkit-keyframes k1 {
  from {
    color: red;
  }
  to {
    color: blue;
  }
}
@-moz-keyframes k1 {
  from {
    color: red;
  }
  to {
    color: blue;
  }
}
@keyframes k1 {
  from {
    color: red;
  }
  to {
    color: blue;
  }
}
.a {
  animation-name: k1;
}
.b {
  animation-duration: 5s;
}


<div className=ui-provider dir=ltr>
  <div className=ui-box ui-animation a b />
</div>;
"
`;

exports[`felaRenderer marginLeft is rendered into marginLeft due to RTL with \`noFlip\` 1`] = `
".a {
  margin-left: 10px; /* @noflip */
}


<div className=ui-provider dir=rtl>
  <span className=ui-text a dir=auto>
    Hello
  </span>
</div>;
"
`;

exports[`felaRenderer marginLeft is rendered into marginRight due to RTL 1`] = `
".a {
  margin-right: 10px;
}


<div className=ui-provider dir=rtl>
  <span className=ui-text a dir=auto>
    Hello
  </span>
</div>;
"
`;

exports[`felaRenderer styles are expanded to longhand values 1`] = `
".a {
  border-top-style: solid;
}
.b {
  border-right-style: solid;
}
.c {
  border-bottom-style: solid;
}
.d {
  border-left-style: solid;
}
.e {
  border-top-color: rgba(51, 204, 51, 1);
}
.f {
  border-right-color: rgba(51, 0, 204, 1);
}
.g {
  border-bottom-color: rgba(51, 204, 51, 1);
}
.h {
  border-left-color: rgba(51, 0, 204, 1);
}


<div className=ui-box a b c d e f g h />;
"
`;
